plugins/node/opentelemetry-koa-instrumentation/src/koa.ts

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin } from '@opentelemetry/core';
+import * as koa from 'koa';
+import * as shimmer from 'shimmer';
+import {
+  Parameters,
+  KoaMiddleware,
+  KoaContext,
+  KoaComponentName,
+  kLayerPatched,
+} from './types';
+import { VERSION } from './version';
+import { getMiddlewareMetadata } from './utils';
+
+/** Koa instrumentation for OpenTelemetry */
+export class KoaInstrumentation extends BasePlugin<typeof koa> {
+  static readonly component = KoaComponentName;
+  readonly supportedVersions = ['^2.0.0'];
+
+  constructor(readonly moduleName: string) {
+    super('@opentelemetry/koa-instrumentation', VERSION);
+  }
+
+  /**
+   * Patches Koa operations by wrapping the Koa.use function
+   */
+  protected patch(): typeof koa {
+    this._logger.debug('Patching Koa');
+
+    if (this._moduleExports === undefined || this._moduleExports === null) {
+      return this._moduleExports;
+    }
+    this._logger.debug('Patching Koa.use');
+    const appProto = this._moduleExports.prototype;
+    shimmer.wrap(appProto, 'use', this._getKoaUsePatch.bind(this));
+
+    return this._moduleExports;
+  }
+
+  /**
+   * Unpatches all Koa operations
+   */
+  protected unpatch(): void {
+    const appProto = this._moduleExports.prototype;
+    shimmer.unwrap(appProto, 'use');
+  }
+
+  /**
+   * Patches the Koa.use function in order to instrument each original
+   * middleware layer which is introduced
+   * @param {KoaMiddleware} middleware - the original middleware function
+   */
+  private _getKoaUsePatch(original: (middleware: KoaMiddleware) => koa) {
+    return function use(
+      this: koa,
+      middlewareFunction: KoaMiddleware,
+      ...args: Parameters<typeof original>
+    ) {
+      let patchedFunction;
+      if (middlewareFunction.router) {
+        patchedFunction = plugin._patchRouterDispatch(middlewareFunction);
+      } else {
+        patchedFunction = plugin._patchLayer(middlewareFunction, false);
+      }
+
+      args[0] = patchedFunction;
```

Comment:
```suggestion
      args[0] = middlewareFunction.router ?
        plugin._patchRouterDispatch(middlewareFunction) :
        patchedFunction = plugin._patchLayer(middlewareFunction, false);
```

Reply:
Why are you overwriting `args[0]`? Did you mean to use something like `unshift` instead?

Reply:
To avoid overwriting `args`, you could also just do 
```ts
const patchedFunction = middlewareFunction.router ?
        plugin._patchRouterDispatch(middlewareFunction) :
        patchedFunction = plugin._patchLayer(middlewareFunction, false);
return original.apply(this, [ patchedFunction, ...args ])
```

Reply:
Thanks for the suggestions! I ended up removing args entirely because the Koa `app.use` function does not allow any additional parameters anyways.